datastore/sql/sqliteds.nim

```diff
@@ -308,7 +307,7 @@ method query*(
   iter.dispose = proc(): Future[?!void] {.async.} =
     discard sqlite3_reset(s)
     discard sqlite3_clear_bindings(s)
-    s.dispose
+    iter.next = nil
```

Comment:
I think we'll still need to do a `dispose` on the query statement. Otherwise it looks like it'll leak the query statements. 

If this dispose is causing the issue then possibly we're calling the iterator's dispose multiple times? I've been meaning to create a PR to pull out some of the fixes I did in the threaded datastore for how codex uses the query iterator. Although most of those issues were due to Codex not calling dispose on the iterator, IIRC. 

Possibly we might need to add `isFinished` flag to ensure we only call dispose once? Alternatively a destructor for QueryIter could work.

Reply:
> I think we'll still need to do a dispose on the query statement. Otherwise it looks like it'll leak the query statements.

`dispose` just calls `sqlite3_reset(s)` and sets the statement to `nil`, it's not required here because we're already calling `sqlite3_reset` on line 308.

Reply:
It looks like here `dispose` is calling `sqlite3_finalize` function not `sqlite3_reset`. 

https://github.com/codex-storage/nim-datastore/blob/d02f2e7380410137fd2cdc9bd0d06c0004f7e8ff/datastore/sql/sqliteutils.nim#L92

Also the sqlite3 [docs](https://www.sqlite.org/c3ref/finalize.html) say `sqlite3_reset` is just to reset the query state, not to clean it up:

> The sqlite3_finalize(S) routine can be called at any point during the life cycle of [prepared statement](https://www.sqlite.org/c3ref/stmt.html) S: before statement S is ever evaluated, after one or more calls to [sqlite3_reset()](https://www.sqlite.org/c3ref/reset.html), or after any call to [sqlite3_step()](https://www.sqlite.org/c3ref/step.html) regardless of whether or not the statement has completed execution.

Reply:
@dryajov just pinging you to make sure you saw my comment above.

Reply:
> It looks like here `dispose` for the `RawStmtPtr` is calling `sqlite3_finalize` function not `sqlite3_reset`.
> 
> https://github.com/codex-storage/nim-datastore/blob/d02f2e7380410137fd2cdc9bd0d06c0004f7e8ff/datastore/sql/sqliteutils.nim#L92
> 
> Also the sqlite3 [docs](https://www.sqlite.org/c3ref/finalize.html) say `sqlite3_reset` is just to reset the query state, not to clean it up:
> 
> > The sqlite3_finalize(S) routine can be called at any point during the life cycle of [prepared statement](https://www.sqlite.org/c3ref/stmt.html) S: before statement S is ever evaluated, after one or more calls to [sqlite3_reset()](https://www.sqlite.org/c3ref/reset.html), or after any call to [sqlite3_step()](https://www.sqlite.org/c3ref/step.html) regardless of whether or not the statement has completed execution.

Ah, you might be right, I'll take another look.